lib/server/env/Translation.js

```diff
@@ -0,0 +1,190 @@
+        // for pages we have more strict rules
+        throw new Error(
+          `Text that you've identified for translation ('${key}') hasn't been added to the global list in 'en.json'. To solve this problem run 'yarn write-translations'.`
+        );
+      }
+    } else {
+      return this.t(key, category, {
+        language: DEFAULT_LANG,
+        fallbackFrom: language,
+      });
+    }
+
+    return null;
+  }
+
+  enabledLanguages = () => this.languages.filter(lang => lang.enabled);
+
+  /**
+   * Parse locale from doc file path
+   *
+   * @param  {string} file
+   * @param  {string} refDir
+   *
+   * @return {?string}
+   */
+  getFileLanguage(file, refDir) {
+    const separator = escapeStringRegexp(path.sep);
+    const baseDir = escapeStringRegexp(path.basename(refDir));
+    const regexSubFolder = new RegExp(
+      `${baseDir}${separator}(.*?)${separator}.*`
+    );
+    const match = regexSubFolder.exec(file);
+
+    // Avoid misinterpreting subdirectory as language
+    if (match && this.enabled) {
+      const enabledLanguages = this.enabledLanguages().map(
+        language => language.tag
+      );
+
+      if (enabledLanguages.indexOf(match[1]) !== -1) {
+        return match[1];
+      }
+    }
+
+    return null;
+  }
+
+  load() {
+    if (fs.existsSync(languagesFile)) {
+      this.enabled = true;
+      this.languages = require(languagesFile);
+    }
+
+    this.loadTranslations();
+  }
+
+  loadTranslations() {
+    const translations = {languages: this.enabledLanguages()};
+
+    const files = glob.sync(`${CWD}/i18n/**`);
+    const langRegex = /\/i18n\/(.*)\.json$/;
+    let wasWarned = false;
+
+    files.forEach(file => {
+      const extension = path.extname(file);
+
+      if (extension === '.json') {
+        const match = langRegex.exec(file);
+        const language = match[1];
+
+        translations[language] = require(file);
+
+        Object.defineProperty(this, language, {
+          get() {
+            if (!wasWarned) {
+              wasWarned = true;
+
+              console.error(
+                "translation[language][category][key] api is deprecated. Please use translation.t(key, category = 'default')"
+              );
+            }
+
+            return translations[language];
+          },
+          set() {
+            // noop
+          },
+        });
+      }
+    });
+
+    this.translations = translations;
+  }
+
+  /* handle escaped characters that get converted into json strings */
+  parseEscapeSequences(str) {
+    return str
+      .replace(new RegExp('\\\\n', 'g'), '\n')
+      .replace(new RegExp('\\\\b', 'g'), '\b')
+      .replace(new RegExp('\\\\f', 'g'), '\f')
+      .replace(new RegExp('\\\\r', 'g'), '\r')
+      .replace(new RegExp('\\\\t', 'g'), '\t')
+      .replace(new RegExp("\\\\'", 'g'), "'")
+      .replace(new RegExp('\\\\"', 'g'), '"')
+      .replace(new RegExp('\\\\', 'g'), '\\');
+  }
+
+  resolveMessage(translations, keyParts) {
```

Comment:
I think this is too generic to be put inside translation.

We should preserve the safe-getter `idx`  function in `core/utils.js` to access deeply nested object and this class can instead import that utility function

The `resolveMessage` in translation just need to require this utils function